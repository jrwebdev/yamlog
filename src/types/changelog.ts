export type ChangeType = 'breaking' | 'feature' | 'fix';

export interface DetailedChange {
  details: string;
  module?: string;
}

export type Change = string | DetailedChange;

export interface ChangelogVersionMetadata {
  date: string;
}

type ChangelogVersionChanges = { [type in ChangeType]?: Change[] };

export interface ChangelogVersion extends ChangelogVersionChanges {
  metadata?: ChangelogVersionMetadata;
}

export interface Changelog {
  [version: string]: ChangelogVersion;
}
