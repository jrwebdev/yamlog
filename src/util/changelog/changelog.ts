import { compose } from 'ramda';
import {
  ChangeType,
  Change,
  DetailedChange,
  ChangelogVersion,
} from './changelog.types';

import * as unreleasedDir from './unreleased-dir';

const changeTypeKeys: ChangeType[] = ['breaking', 'feature', 'fix'];

const isDetailedChange = (change: any): change is DetailedChange =>
  typeof change === 'object';

const stripEmptyChangeTypes = (version: ChangelogVersion) =>
  changeTypeKeys.reduce(
    (acc, key) => {
      // TODO: Fix type
      if (acc[key] && !(acc as any)[key].length) {
        delete acc[key];
      }
      return acc;
    },
    { ...version }
  );

const stripEmptyChangeProperties = (change: Change): DetailedChange | Change =>
  isDetailedChange(change)
    ? Object.keys(change).reduce((acc, property) => {
        const value = (change as any)[property];
        if (value && value.trim()) {
          (acc as any)[property] = value;
        }
        return acc;
      }, {}) as DetailedChange
    : change;

const flattenDetails = (change: Change) =>
  typeof change !== 'string' &&
  Object.keys(change).length === 1 &&
  change.details
    ? change.details
    : change;

const processChange = compose(flattenDetails, stripEmptyChangeProperties);

export const addChange = (changeType: ChangeType, change: Change) => {
  const processedChange = processChange(change);
  unreleasedDir.addChange(changeType, processedChange);
};
