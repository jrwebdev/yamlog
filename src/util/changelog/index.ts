export * from './changelog';
export * from './changelog.types';
