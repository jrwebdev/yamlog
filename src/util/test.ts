export const getMock = <T>(fn: (...args: any[]) => T) => fn as jest.Mock<T>;
