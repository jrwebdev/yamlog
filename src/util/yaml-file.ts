import { readFile, writeFile } from 'fs-extra';
import { parse, format } from './yaml';

export const read = async (filename: string) =>
  parse((await readFile(filename)).toString());

export const write = (filename: string, data: any) =>
  writeFile(filename, format(data));
