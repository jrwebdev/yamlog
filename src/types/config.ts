import { Changelog } from './changelog';
import { VersionString } from './version';

type CurrentVersionLoaderFn = (
  changelog?: Changelog
) => Promise<VersionString | undefined> | VersionString | undefined;

export type CurrentVersionLoader =
  | CurrentVersionLoaderFn
  | CurrentVersionLoaderFn[];

export interface ChangelogConfig {
  unstable?: boolean;
  startVersion?: string;
  // currentVersionLoader?: CurrentVersionLoader;
}

export default interface Config extends ChangelogConfig {
  commit?: {
    cmd: string;
  };
  publish?: {
    bump?: false;
    commit?: false;
    tag?: false;
    push?: false;
    dir?: string;
    copyFiles?: string[] | false;
    pre?: string | string[];
    post?: string | string[];
  };
};
