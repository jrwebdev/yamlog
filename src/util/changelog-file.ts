import { existsSync } from 'fs-extra';
import { ChangeType, Change, Changelog } from '../types/changelog';
import { read as readYamlFile, write as writeYamlFile } from './yaml-file';

const changelogFile = 'changelog.yaml';

const read = async () =>
  existsSync(changelogFile)
    ? (await readYamlFile(changelogFile)) as Changelog
    : {};

const write = (changelog: Changelog) => writeYamlFile(changelogFile, changelog);

// TODO: Sort changes
export const addChange = async (changeType: ChangeType, change: Change) => {
  const { unreleased = {}, ...released } = await read();
  unreleased[changeType] = [change, ...(unreleased[changeType] || [])];
  return write({ unreleased, ...released });
};
