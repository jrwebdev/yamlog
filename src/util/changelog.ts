import * as moment from 'moment';

import { ChangelogConfig, CurrentVersionLoader } from '../types/config';
import { Changelog, ChangeType, Change } from '../types/changelog';

import {
  addChange as addUnreleasedDirChange,
  read as readUnreleasedFiles,
  removeDir as deleteUnreleasedFiles,
} from './unreleased-dir';
import { getNextVersion } from './changelog-helpers';
import {
  addChange as addChangelogChange,
  read as readChangelog,
  write as writeChangelog,
} from './changelog-file';

const loadCurrentVersion = async (
  loaders: CurrentVersionLoader,
  releasedVersions: Changelog
) => {
  if (Array.isArray(loaders)) {
    let version;
    let i = 0;
    // TODO: Change to use recursion
    while (!version && loaders[i]) {
      version = await loaders[i](releasedVersions);
      i += 1;
    }
    return version;
  } else {
    return await loaders(releasedVersions);
  }
};

export const addChange = (
  changeType: ChangeType,
  change: Change,
  { unreleasedDir = '' } = {}
) =>
  unreleasedDir
    ? addUnreleasedDirChange(changeType, change, unreleasedDir)
    : addChangelogChange(changeType, change);

export const bumpVersion = async (
  currentVersionLoader: CurrentVersionLoader,
  config: ChangelogConfig = {}
) => {
  const {
    unreleased: changelogUnreleased,
    ...released,
  } = await readChangelog();

  // TODO: Combine?
  const unreleased = config.unreleasedDir
    ? await readUnreleasedFiles(config.unreleasedDir)
    : changelogUnreleased;

  if (unreleased && Object.keys(unreleased).length) {
    const currentVersion = await loadCurrentVersion(
      currentVersionLoader,
      released
    );

    const newVersion = getNextVersion(unreleased, currentVersion, config);
    if (newVersion) {
      await writeChangelog({
        [newVersion]: {
          metadata: {
            date: moment().format('YYYY-MM-DD'),
          },
          ...unreleased,
        },
        ...released,
      });

      if (config.unreleasedDir) {
        await deleteUnreleasedFiles(config.unreleasedDir);
      }

      return newVersion;
    }
  }

  return undefined;
};
