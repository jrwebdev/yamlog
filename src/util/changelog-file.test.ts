import { Changelog } from '../types/changelog';
import { read, write } from './yaml-file';

import { getMock } from './test';

import { addChange } from './changelog-file';

jest.mock('fs-extra', () => ({
  existsSync: () => true,
}));

jest.mock('./yaml-file');

let mockChangelog: Changelog;

const getOutput = () => getMock(write).mock.calls[0][1];

beforeEach(() => {
  mockChangelog = {
    unreleased: {
      breaking: ['a breaking change'],
      feature: ['a feature'],
      fix: ['a fix'],
    },
    '0.1.0': {
      metadata: {
        date: '2017-01-01',
      },
      feature: ['a feature'],
    },
  };

  getMock(read).mockImplementation(() => Promise.resolve(mockChangelog));
  getMock(write).mockImplementation(() => Promise.resolve());

  expect.hasAssertions();
});

afterEach(() => {
  jest.resetAllMocks();
});

describe('addChange', () => {
  it('should write a change to the changelog.yaml file', async () => {
    await addChange('feature', '** NEW CHANGE **');
    expect(write).toHaveBeenCalledWith('changelog.yaml', jasmine.anything());
  });

  it('should write a change to the start of the breaking section', async () => {
    await addChange('breaking', '** NEW CHANGE **');
    expect(getOutput()).toMatchSnapshot();
  });

  it('should write a change to the start of the feature section', async () => {
    await addChange('feature', '** NEW CHANGE **');
    expect(getOutput()).toMatchSnapshot();
  });

  it('should write a change to the start of the fix section', async () => {
    await addChange('fix', '** NEW CHANGE **');
    expect(getOutput()).toMatchSnapshot();
  });

  it('should create a new unreleased section if one does not exist', async () => {
    delete mockChangelog.unreleased;
    await addChange('feature', '** NEW CHANGE **');
    expect(getOutput()).toMatchSnapshot();
  });

  it('should create a new change type section if one does not exist', async () => {
    delete mockChangelog.unreleased.feature;
    await addChange('feature', '** NEW CHANGE **');
    expect(getOutput()).toMatchSnapshot();
  });

  it('should strip out any empty sections', async () => {
    mockChangelog.unreleased = {
      breaking: [],
      feature: [],
      fix: mockChangelog.unreleased.fix,
    };
    await addChange('fix', '** NEW CHANGE **');
    expect(getOutput()).toMatchSnapshot();
  });

  it('should reorganise the change types if they are out of order', async () => {
    mockChangelog.unreleased = {
      fix: mockChangelog.unreleased.fix,
      breaking: mockChangelog.unreleased.breaking,
      feature: mockChangelog.unreleased.feature,
    };
    await addChange('fix', '** NEW CHANGE **');
    expect(getOutput()).toMatchSnapshot();
  });

  it('should output a detailed changelog entry', async () => {
    await addChange('feature', {
      details: '** NEW CHANGE **',
      module: 'module',
    });
    expect(getOutput()).toMatchSnapshot();
  });

  it('should add the change as a string if the detailed changelog entry only contains details', async () => {
    await addChange('feature', { details: '** NEW CHANGE **' });
    expect(getOutput()).toMatchSnapshot();
  });
});
