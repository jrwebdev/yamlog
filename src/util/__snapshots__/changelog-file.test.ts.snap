// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addChange should add the change as a string if the detailed changelog entry only contains details 1`] = `
unreleased:
  breaking:
    - a breaking change
  feature:
    - '** NEW CHANGE **'
    - a feature
  fix:
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should create a new change type section if one does not exist 1`] = `
unreleased:
  breaking:
    - a breaking change
  feature:
    - '** NEW CHANGE **'
  fix:
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should create a new unreleased section if one does not exist 1`] = `
unreleased:
  feature:
    - '** NEW CHANGE **'
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should output a detailed changelog entry 1`] = `
unreleased:
  breaking:
    - a breaking change
  feature:
    - details: '** NEW CHANGE **'
      module: module
    - a feature
  fix:
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should reorganise the change types if they are out of order 1`] = `
unreleased:
  breaking:
    - a breaking change
  feature:
    - a feature
  fix:
    - '** NEW CHANGE **'
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should strip out any empty sections 1`] = `
unreleased:
  fix:
    - '** NEW CHANGE **'
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should write a change to the start of the breaking section 1`] = `
unreleased:
  breaking:
    - '** NEW CHANGE **'
    - a breaking change
  feature:
    - a feature
  fix:
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should write a change to the start of the feature section 1`] = `
unreleased:
  breaking:
    - a breaking change
  feature:
    - '** NEW CHANGE **'
    - a feature
  fix:
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;

exports[`addChange should write a change to the start of the fix section 1`] = `
unreleased:
  breaking:
    - a breaking change
  feature:
    - a feature
  fix:
    - '** NEW CHANGE **'
    - a fix
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - a feature

`;
